### PR TITLE
BF: HTTP Downloader treated 429 as authentication issue

### LINKED ISSUE #7129: test_ria_basics.py::test_binary_data: wants interaction
Prior issue (closed as addressed) for this test was #4371 , suggests something about Authorization (didn't check in detail) hence might indeed relate to the desire of interaction.

Happened in the course of testing https://github.com/datalad/datalad/pull/7126 which AFAIK does not relate, happened on travis in a single run of the heavy debug logging etc:

```
=========================== short test summary info ============================
FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully. 1 failed:
[{'action': 'download_url',
  'error_message': "'ConsoleLog' object has no attribute 'yesno'",
  'exception': 'ConsoleLog' object has no attribute 'yesno' [download_url.py:__call__:202,base.py:download:527,base.py:access:219,base.py:_handle_authentication:286,base.py:_enter_credentials:319,__init__.py:__getattribute__:101,base.py:access:174,base.py:_download:427,http.py:get_downloader_session:622,http.py:check_response_status:124],
  'exception_traceback': '[download_url.py:__call__:202,base.py:download:527,base.py:access:219,base.py:_handle_authentication:286,base.py:_enter_credentials:319,__init__.py:__getattribute__:101,base.py:access:174,base.py:_download:427,http.py:get_downloader_session:622,http.py:check_response_status:124]',
  'message': "AttributeError('ConsoleLog' object has no attribute 'yesno')",
  'path': '/tmp/datalad_temp__test_binary_datathev6l40/dicomfile',
  'status': 'error',
  'type': 'file'}]
= 1 failed, 1162 passed, 46 skipped, 3 xfailed, 17 warnings in 1023.67s (0:17:03) =
The command "http_proxy= PATH=$PWD/../tools/coverage-bin:$PATH $PYTEST_WRAPPER python -m pytest "${PYTEST_OPTS[@]}" -c ../tox.ini -n 2 -m "${PYTEST_SELECTION:+$PYTEST_SELECTION_OP($PYTEST_SELECTION) and }not(turtle)" --doctest-modules --cov=datalad --cov-report= --pyargs $TESTS_TO_PERFORM" exited with 1.
0.00s$ cd ..
```

Seems to be relatively frequent encounter, and only on travis (looked only at this month):

```
(git)smaug:/mnt/datasets/datalad/ci/logs/2022/10[master]git
$> git grep '^FAILED.*test_binary_data'
11/pr/7068/08876f3/travis-14700-failed/9.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
12/pr/7002/16a162d/travis-14714-failed/8.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
19/pr/7094/a88c6e1/travis-14761-failed/2.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
19/push/master/1de6d08/travis-14764-failed/2.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
21/push/maint/e94a49e/travis-14772-failed/4.txt:FAILED ../distributed/tests/test_ria_basics.py::test_binary_data - datalad.su...
24/push/maint/969bb82/travis-14794-failed/9.txt:FAILED ../datalad/distributed/tests/test_ria_basics.py::test_binary_data - da...
git grep '^FAILED.*test_binary_data'  4.04s user 3.52s system 1016% cpu 0.743 total
```
and seems to happen not only in the "heavy debug" run

### PR DESCRIPTION
**Pull request by bpoldrack**
Fixes #7129

### CODE REVIEW

File: datalad/downloaders/base.py [hunk 1]
```diff
@@ -176,6 +176,12 @@ def access(self, method, url, allow_old_session=True, **kwargs):
                 break
             except AccessDeniedError as e:
                 ce = CapturedException(e)
+                if hasattr(e, 'status') and e.status == 429:
```
Comment (yarikoptic) [inline datalad/downloaders/base.py:179]: and for `429` we get AccessDeniedError and not DownloadError which is handled below?  
do we have logging when it happens or we might want to add an explicit `lgr.debug("` while treating it with sleep

File: datalad/downloaders/base.py [hunk 2]
```diff
@@ -176,6 +176,12 @@ def access(self, method, url, allow_old_session=True, **kwargs):
...
+                    # Too many requests.
+                    # We can retry by continuing the loop.
+                    time.sleep(0.5)
```
Comment (yarikoptic) [inline datalad/downloaders/base.py:182]:
```suggestion
                    time.sleep(0.5*(attempt**1.2))
```
some exponential growth is good here ;)

File: datalad/downloaders/base.py (hunk 1 above)
Comment (bpoldrack) [reply datalad/downloaders/base.py:179]: > and for 429 we get AccessDeniedError and not DownloadError which is handled below?

Yes. We raise `AccessDeniedError` for any code in 400..500 range. And to some extend the notion of "denied" is alright. It's just that this part here regards that as authentication related which it is not.

Comment (bpoldrack) [reply datalad/downloaders/base.py:179]: > do we have logging when it happens 

`CapturedException` logs the exception and there's level 5 log on every attempt. I'd prefer to not add more. In my view it's already hard enough to find the relevant messages in the logs. If everything is repeated and rephrased at higher log levels, things get just harder to read, I think.

File: changelog.d/pr-7142.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+### 🐛 Bug Fixes
+
+- HTTP download treated '429 - too many requests' as an authentication issue and
+  was consequently trying to obtain credentials.
+  Fixes [#7129](https://github.com/datalad/datalad/issues/7129) via
+  [PR #YYY](https://github.com/datalad/datalad/pull/YYY)
```
Comment (yarikoptic) [inline changelog.d/pr-7142.md:6]:
```suggestion
  [PR #7129](https://github.com/datalad/datalad/pull/7129)
```
we need to change default to `maint`... will do now

File: datalad/downloaders/base.py (hunk 1 above)
Comment (yarikoptic) [reply datalad/downloaders/base.py:179]: ok

**Review by yarikoptic — APPROVED** [review]

**Comment by yarikoptic-gitmate** [comment]
PR released in [`0.9.4`](https://github.com/datalad/datalad/releases/tag/0.9.4)

**Comment by yarikoptic-gitmate** [comment]
PR released in [`0.17.9`](https://github.com/datalad/datalad/releases/tag/0.17.9)

**Pull request merged**